MAINT: backports for SciPy 1.16.3

tylerjereddy opened the pull request:
There's a fairly credible request for SciPy `1.16.3` at https://github.com/scipy/scipy/pull/23787#issuecomment-3413210944. That said, let's first test the waters here with a regular flush of the CI to see what kinds of shenanigans we'd have to deal with to make this happen. Also, cc @h-vetinari and @jorenham for planning visibility.

Backports included here (so far):
1. gh-23681
2. gh-23685
3. gh-23777
4. gh-23787
5. gh-23829
6. gh-23821
7. gh-23817
8. gh-23815
9. gh-23864
10. (partial backport only) gh-23811

Backports included prior this PR:
1. gh-23614 (was backported in gh-23615)
2. gh-23629 (was backported in gh-23615)


TODO list:

- [x] assess the situation with regular CI--tractable for a release without too much pain?
- [x] full CI passing, including wheel builds
- [x] update the `1.16.3` release notes/author lists, etc.
- [x] any other backports we need/want to wait a bit for?
- [x] ~should we pin `array-api-strict<2.4.0` in the various dependency specification files?~

rgommers committed:
MAINT: optimize: fix usage of `nextafter(..., where=)` without `out=`. (#23685)

This raises a warning now (after numpy#29813) because it returns uninitialized
memory. The usage seems correct, so I left the code as is, although the
performance difference of using `where=` over `np.where` is probably
negligible.

Fixes CI failures in jobs using numpy nightlies that look like:
```
 FAILED scipy/optimize/tests/test_minimize_constrained.py::test_gh11649 - UserWarning: 'where' used without 'out', expect unitialized memory in output. If this is intentional, use out=None.
```

mdhaber committed:
BUG: linalg:solve...are: fix bug when optional array is skipped (#23681)

* MAINT: fix solve_*_are when optional array is skipped

* MAINT: linalg: pass Nones through _apply_over_batch

nickodell committed:
CI: pin Python 3.11.13 in CircleCI job (#23777)

ilayn committed:
BUG:optimize:SLSQP: Use double negative in the if condition for NaN handling

mdhaber committed:
TST: linalg.svd: skip on 32-bit due to limited memory

mdhaber committed:
ENH: scipy.stats: speed up bivariate normal cdf

tylerjereddy commented:
The one CI failure with `torch` is gh-23820--since that's just a regex mismatch on an expected error message after the `torch 2.9.0` update this week, I suspect any workaround to get that passing should be fine.

nickodell committed:
CI: optimize: allow new Pytorch message

Pytorch 2.9.0 changes the error message on failed broadcasts. Allow
the new style

ChrisAB committed:
BUG/TST: Bump xsf to v0.1.3 add tests for expi limits (#23817)

* TST: Adds test for limits for expi function from xsf\

Requires closure of https://github.com/scipy/xsf/pull/66 which closes #23738

* Bumped xsf to v0.1.3
Closes #23738

* Refactor test_limits to use parameterized tests

---------

Co-authored-by: Chrisab <ChrisAB>

mdhaber committed:
BLD: stats.multivariate_normal: suppress pythran build warning

mdhaber committed:
STY: stats.multivariate_normal: add license note

mdhaber committed:
Revert "BLD: stats.multivariate_normal: suppress pythran build warning"

This reverts commit 4b7dcc1faf2740e47853bec536e900a6c6502903.

mdhaber committed:
BLD: stats.multivariate_normal: silence pythran build warning

rlucas7 committed:
MAINT: patch vendored qhull source for security fix (#23829)

Co-authored-by: Lucas Roberts <rlucas7@Mac.lan>
Co-authored-by: Lucas Colley <lucas.colley8@gmail.com>

tylerjereddy committed:
DOC: PR 23805 revisions

* Update the SciPy `1.16.3` release notes
following additional backport activity.

tylerjereddy commented:
Ok, we're getting closer to a "critical mass" of backports now, we're up to 10 after the latest update. The one labelled "ENH:" is arguably a bug fix for a severe performance regression.

Local testing on ARM Mac seems "ok" (ignoring the usual flakes I get), including array API tests and preview of the new release notes from (local) doc build.

Note that we need `array-api-strict<2.4.0` on the release branch for array API tests to pass, and that was true prior to this release attempt. While that version limitation is already set in CI, it is *not* set in `pyproject.toml` nor in the variety of other files that we can use to pull in/specify deps. I haven't changed that, not sure if it matters a whole lot, but I'll add it to the TODO list above I suppose. I'm somewhat-tempted to avoid touching dep specifications for an optional testing package, though on the other hand it should be harmless/slightly helpful I suppose.

Anyway, let's see if the regular CI is finally green now.

jorenham commented on doc/source/release/1.16.3-notes.rst:
```diff
@@ -12,12 +12,37 @@ compared to 1.16.2.
 
+* `#23336 <https://github.com/scipy/scipy/issues/23336>`__: BUG: linalg: wrong handling of args/kwargs in ``_apply_over_batch``
+* `#23469 <https://github.com/scipy/scipy/issues/23469>`__: BUG: huge speed and quality regression for stats.multivariate_normal.cdf
+* `#23642 <https://github.com/scipy/scipy/issues/23642>`__: CI: meson setup fails with "invalid int value: 'Sentinel.UNSET'"
```
I see that xsf has been updated, which means that the fix (https://github.com/scipy/xsf/pull/62) for #23589 is also included:

```suggestion
* `#23589 <https://github.com/scipy/scipy/issues/23589>`__: BUG: ``lqmn(1,1,0)`` and  ``lqmn(1,1,0j)`` return incompatible derivatives
* `#23642 <https://github.com/scipy/scipy/issues/23642>`__: CI: meson setup fails with "invalid int value: 'Sentinel.UNSET'"
```

tylerjereddy replied:
Oh thanks--it didn't get slurped up by the automated release notes list scripts because it was lacking a milestone (I just added the `1.16.3` milestone over there).

That said, this does make me think about something else--I'm pretty sure the author list/name script won't pick you up as an author on the release since your patch was in a different repo.

My initial inclination is to be reasonably inclusive on authorship, maybe I can just add you manually in this case. Is there a feeling that `xsf` is sufficiently independent that folks working on it are happy to get author "credit" over there on its own releases? I don't have a good read on that--I suppose `xsf` has prospects for wide usage separate from SciPy anyway, so maybe?

rgommers replied:
I would say that the primary visibility of authorship is commit counts, and `xsf` has its own (https://github.com/scipy/xsf/graphs/contributors). So I wouldn't do work to treat `xsf` differently from all other libraries we vendor.

jorenham replied:
I was mostly talking about the resolved issue I raised in the scipy repo, actually

tylerjereddy commented:
Of course there are now more CI failures than before. Spending time on pre-release failures is perhaps a bit dubious though I've generally tried to investigate them. I'm guessing the doc build warnings are my fault somehow.

mdhaber committed:
TST: stats.boxcox_llf: bump test tolerance for Accelerate (#23864)

tylerjereddy committed:
CI: PR 23805 revisions

* Pin the version of Cython used by the Linux free-threaded
CI job to avoid pulling in an alpha/beta `3.2.x` release, as
noted at:
https://github.com/scipy/scipy/pull/23805#issuecomment-3453068689

tylerjereddy committed:
DOC: PR 23805 revisions

* Update the SciPy `1.16.3` release notes following
additional backport activity in the above PR.

tylerjereddy committed:
MAINT: PR 23805 wheel builds [wheel build]

* An empty commit to test the full wheel
build matrix ahead of the SciPy `1.16.3`
release.

[wheel build]

tylerjereddy commented:
all green, in it goes

tylerjereddy commented:
@rlucas7 note the "immutability" marker we talked about with the new security setting on releases:

<img width="206" height="305" alt="image" src="https://github.com/user-attachments/assets/4306c0fc-4e1c-4040-b44c-490245960774" />

It does let me edit the text of the release, but NOT the tag or assets--I suppose that's fine.
